### DISCUSSION TITLE
[Proposal] MathExpressionConverter boolean support

### DISCUSSION THREAD

**Discussion by stephenquan** [CommunityToolkit/Maui#2084] [New Feature Discussions]
This is a Math Expression Converters https://github.com/CommunityToolkit/Maui/issues/71 refactor enhancement to improve the parser by (1) using a Parsing expression grammar, and (2) to support boolean operations. The grammar is now:

```
Expr        ::= Conditional
Conditional ::= LogicalOr ( '?' LogicalOr ':' LogicalOr ) ?
LogicalOr   ::= LogicalAnd ( '||' LogicalAnd ) *
LogicalAnd  ::= Equality ( '&&' Equality ) *
Equality    ::= Compare ( ( '==' | '!=' ) Compare ) *
Compare     ::= Sum ( ( '>=' | '>' | '<=' | '<' ) Sum ) *
Sum         ::= Product ( ( '+' | '-' ) Product ) *
Product     ::= Power ( ( '*' | '/' ) Power ) *
Power       ::= Primary ( '^' Primary ) *
Primary     ::= Number
              | Function
              | Constant
              | '(' Expr ')'
              | ( '-' | '!' ) Primary
Number      ::= ( '-' ) ? ( ( [0-9]+ '.' [0-9]+ ) | [0-9]+ )
Function    ::= [a-z]+ ( '(' ')' | '(' Expr ( ',' Expr ) * ')' )
Constant    ::= [a-z]+
```

Where `Conditional`, `LogicalOr`, `LogicalAnd`, `Equality`, and `Compare` are new. `true` and `false` boolean constants are also new. `Sum`, `Product`, `Power`, `Primary`, `Number`, `Function`, and `Constant` are a refactor of what was already supported by the original parser.

The implementation can be found:
 - in my fork: https://github.com/stephenquan/Maui
 - preview change commits: https://github.com/CommunityToolkit/Maui/compare/main...stephenquan:Maui:main

References:
 - https://en.wikipedia.org/wiki/Parsing_expression_grammar
 - https://en.wikipedia.org/wiki/Order_of_operations#Programming_languages

**Comment by stephenquan** [comment]
For instance, if you wanted to implement the conditional `Label.TextColor = x0 + x1 + x2 >= 60 ? Colors.Green : Colors.Red` you can make use of the `MultiMathExpression` and write it as follows:

```xaml
<Label.TextColor>
    <MultiBinding Converter="{StaticResource MultiMathExpressionConverter}" ConverterParameter="x0 + x1 + x2 &gt;= 60 ? x3 : x4">
        <Binding Mode="OneWay" Path="X0" />
        <Binding Mode="OneWay" Path="X1" />
        <Binding Mode="OneWay" Path="X2" />
        <Binding Source="{x:Static Colors.Green}" />
        <Binding Source="{x:Static Colors.Red}" />
    </MultiBinding>
</Label.TextColor>
```

Alternatively, you can solve the same problem using `DataTrigger` by having `Green` as the default color and as `Red` as the non-default. This pattern has merit when we want to define multiple setters for the same trigger.

```xaml
<Label
    Grid.Row="1"
    Grid.Column="7"
    Text="&gt;= 60"
    TextColor="Green"
    VerticalOptions="Center"
    VerticalTextAlignment="Center">
    <Label.Triggers>
        <DataTrigger TargetType="Label">
            <DataTrigger.Binding>
                <MultiBinding Converter="{StaticResource MultiMathExpressionConverter}" ConverterParameter="x0 + x1 + x2 &gt;= 60">
                    <Binding Mode="OneWay" Path="X0" />
                    <Binding Mode="OneWay" Path="X1" />
                    <Binding Mode="OneWay" Path="X2" />
                </MultiBinding>
            </DataTrigger.Binding>
            <DataTrigger.Value><x:Boolean>False</x:Boolean></DataTrigger.Value>
            <Setter Property="Label.TextColor" Value="Red" />
        </DataTrigger>
    </Label.Triggers>
</Label>
```

**Comment by CliffAgius** [comment]
I like this idea and it's got some community support I have tagged to be discussed at this months meetup on Thursday 5th September if your able to attend you could add to the conversation.

**Comment by stephenquan** [comment]
@CliffAgius, thanks, I will try to make it.

**Reply by CliffAgius** [reply]
Thanks for attending the meetup you can read about contributing to the MCT [Here](https://github.com/CommunityToolkit/Maui/wiki/Submitting-a-New-Feature).

So that you have an issue to create a PR against I have converted this discussion into an Issue #2181 for you.

**Comment by CliffAgius** [comment]
This discussion was approved in the [September monthly stand-up](https://www.youtube.com/live/vNf4e5PwcTc) and approved by the community so has been converted into Issue #2181 for the work to begin.

Therefore I am closing this discussion as all work will continue on the Issue for tracking.
